metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+            return (v1.u.v_int32 == v2.u.v_int32) ? 0 : ((v1.u.v_int32 < v2.u.v_int32) ? -1 : 1);
+        case NNE_TYPE_UINT32:
+            return (v1.u.v_uint32 == v2.u.v_uint32) ? 0 : ((v1.u.v_uint32 < v2.u.v_uint32) ? -1 : 1);
+        case NNE_TYPE_STRING:
+            return strcmp(v1.u.v_str, v2.u.v_str);
+    }
+
+    return -2;
+}
+
+static int md_nne_add_json_key_value_int(struct json_object *obj,
+        const char *key, int64_t value)
+{
+    struct json_object *value_obj = NULL;
+
+    value_obj = json_object_new_int64(value);
+    if (!value_obj)
+        return RETVAL_FAILURE;
+
+    json_object_object_add(obj, key, value_obj);
+    return RETVAL_SUCCESS;
+}
+
+static int md_nne_add_json_key_value_string(struct json_object *obj,
+        const char *key, const char *value)
+{
+    struct json_object *value_obj = NULL;
+
+    value_obj = json_object_new_string(value);
+    if (!value_obj)
+        return RETVAL_FAILURE;
+
+    json_object_object_add(obj, key, value_obj);
+    return RETVAL_SUCCESS;
+}
+
+static int md_nne_add_json_key_value_null(struct json_object *obj,
+        const char *key)
+{
+    json_object_object_add(obj, key, NULL);
+    return RETVAL_SUCCESS;
+}
+
+static int md_nne_add_json_key_value(struct json_object *obj,
+        const char* key, struct nne_value value)
+{
+    switch(value.type)
+    {
+        case NNE_TYPE_NULL:
+            return md_nne_add_json_key_value_null(obj, key);
+        case NNE_TYPE_INT8:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_int8);
+        case NNE_TYPE_UINT8:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_uint8);
+        case NNE_TYPE_INT16:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_int16);
+        case NNE_TYPE_UINT16:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_uint16);
+        case NNE_TYPE_INT32:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_int32);
+        case NNE_TYPE_UINT32:
+            return md_nne_add_json_key_value_int(obj, key, value.u.v_uint32);
+        case NNE_TYPE_STRING:
+            return md_nne_add_json_key_value_string(obj, key, value.u.v_str);
+    }
+    return RETVAL_FAILURE;
+}
+
+static void md_nne_send_message(struct md_writer_nne *mwn,
+        struct nne_message* msg)
+{
+    static const char *NNE_MESSAGE_TYPE_STR[] = {
+        "event",
+        "bins-1min"
+    };
+
+    static const char *NNE_MESSAGE_SOURCE_STR[] = {
+        "report",
+        "query"
+    };
+
+    struct json_object* obj = json_object_new_object();
+
+    md_nne_add_json_key_value_string(obj, "type", NNE_MESSAGE_TYPE_STR[msg->type]);
+    md_nne_add_json_key_value_int(obj, "ts", msg->tstamp);
+    md_nne_add_json_key_value_string(obj, "node", msg->node);
+
+    if (msg->mccmnc == 0)
+        md_nne_add_json_key_value_null(obj, "mccmnc");
+    else
+        md_nne_add_json_key_value_int(obj, "mccmnc", msg->mccmnc);
```

Comment:
See earlier comment. Isnt it sufficient for backend to just check if mccmnc is set or not? No need to add NULL value.